### DISCUSSION TITLE
Why does thread sed device Network Attaching always fail

### DISCUSSION THREAD

**Discussion by dengyuxin2** [openthread/openthread#10036] [Q&A]
hardware platform：EFR32MG21
openthread commit：2ce3d3bf0  silicon Gecko SDK 4.1.0
gecko_sdk 4.1.1/4.4.4 When using the chip tool commissioning matter over thread ssed device, it always fails. However, in the same situation, using the chip tool to debug matter over thread ftd device can succeed, and using Raspberry Pi's OTBR can successfully commissioning the same matter over thread sed device. What is the reason for this? This problem has been troubling us for a long time Please help analyze it. Thank you very much for your。Is it related to csl features？ 
thread info：
Active Timestamp: 1
Channel: 25
Channel Mask: 0x07fff800
Ext PAN ID: 7e49aca8d0695302
Mesh Local Prefix: fde2:5bf5:7ef9:cefc::/64
Network Key: 3dd94d9bfae02bf189b5f30827de967f
Network Name: OpenThread-b88b
PAN ID: 0xb88b
PSKc: 851a4a3d4fab46cf66d4983cb7d88a9e
Security Policy: 672 onrc

otbr log:
[MobaXterm_root@OpenWrt_20240416_193523.txt](https://github.com/openthread/openthread/files/14996946/MobaXterm_root%40OpenWrt_20240416_193523.txt)
Packet capture：
[thread设备入网_202404161939.zip](https://github.com/openthread/openthread/files/14996980/thread._202404161939.zip)

**Comment by suveshpratapa** [comment]
Hi, can you upgrade to a newer Silicon Labs SDK? We're currently on 4.4.2 as the latest stable release, and also we've made a lot of platform changes relating to CSL and SSEDs since 4.1.1.

To address your particular SDK question:
1. What baud rate is configured on the RCP / otbr-agent?
2. Can you describe the way you're setting up your SSED example, including CLI and CSL parameters?

I'll take a look at your packet trace but as I said we've fixed a lot of issues since 4.1.1 so your best bet is to upgrade to a newer release.

**Reply by dengyuxin2** [reply]
@suveshpratapa  Thank you for your reply！
 1、baud rate：115200
 2、The sub device is a third-party commercial product purchased online, and I do not know these parameters
I used two sensor devices for testing:
One is Eve's human body sensor
![image](https://github.com/openthread/openthread/assets/158288585/ec97c465-7eb9-4b23-8d93-e4056b5757be)
One is a green meter door magnetic sensor

**Reply by suveshpratapa** [reply]
We recommend using a higher baud rate on OTBR for CSL operations if possible. Can you reconfigure your host and RCP to run at baud rate 460800? Please check the sections "3.2.2.1 UART Baud Rate Settings" and "3.2.4 OTBR Configuration for CSL" in [our application note for OTBR](https://www.silabs.com/documents/public/application-notes/an1256-using-sl-rcp-with-openthread-border-router.pdf).

**Comment by dengyuxin2** [comment]
@suveshpratapa hello We have upgraded to the Silicon latest SDK version 4.4.1 and 4.4.2 of  and tested it using your development board，The test is still the same, but joining the thread network failed, What is the reason for this? Is it our compilation error
 But if only the thread protocol is run, it can be successful in the same environment and device with a single thread protocol
![1713433063516](https://github.com/openthread/openthread/assets/158288585/1451b7be-046a-48e4-adb4-879b8645f3bc)

otbr log + Packet capture：
[otbr_log+Packet_capture.zip](https://github.com/openthread/openthread/files/15023273/otbr_log%2BPacket_capture.zip)

the thread device demo：
![1713435260516](https://github.com/openthread/openthread/assets/158288585/0073f594-f48f-4230-a64f-44a748bd1544)

**Reply by suveshpratapa** [reply]
Please see [above comment](https://github.com/orgs/openthread/discussions/10036#discussioncomment-9156110) for reconfiguring your OTBR installation with the latest SDK.

**Reply by dengyuxin2** [reply]
@suveshpratapa  I set the baud rate to 460800 and 921600 respectively and tested it with the same result. However, when I first added an ftd device and then added a low-power device, the low-power device was added to ftd as its The child device node can successfully connect to the network. If the low-power device directly uses otbr as the parent node, it will always fail. 
I run otbr-agent as follows：
./otbr-agent -I wpan0 -v -B br-lan "spinel+cpc://cpcd_0?iid=2&iid-list=0&cpc-bus-speed=921600"
I run cpcd  as follows：
![image](https://github.com/openthread/openthread/assets/158288585/e3b9095a-aeab-411a-af32-0247be9a34d7)
 Cross-compile otbr-agent settings are as follows：

cmake ../ -DCMAKE_TOOLCHAIN_FILE=../toolchain.cmake -DCMAKE_PREFIX_PATH=/home/dyx/third_lib/lib -DOT_THREAD_VERSION=1.3 -DOT_CHILD_SUPERVISION=ON -DOT_MULTIPAN_RCP=ON -DOT_POSIX_RCP_VENDOR_BUS=CPC -DBUILD_TESTING=OFF -DCMAKE_INSTALL_PREFIX=/usr -DOTBR_DBUS=OFF -DOTBR_DNSSD_DISCOVERY_PROXY=ON -DOTBR_SRP_ADVERTISING_PROXY=ON -DOTBR_INFRA_IF_NAME=br-lan -DOTBR_MDNS=mDNSResponder -DOTBR_VERSION= -DOT_PACKAGE_VERSION= -DOTBR_BORDER_ROUTING=ON -DOTBR_DUA_ROUTING=ON -DOTBR_NO_AUTO_ATTACH=1 -DOT_REFERENCE_DEVICE=ON -DOT_DHCP6_CLIENT=ON -DOT_DHCP6_SERVER=ON -DOTBR_BACKBONE_ROUTER=ON -DCPCD_SOURCE_DIR=/home/dyx/SiliconLabs/cpc-daemon -DOT_POSIX_RCP_VENDOR_BUS=ON -DOT_POSIX_CONFIG_RCP_VENDOR_DEPS_PACKAGE=$GSDK_DIR/protocol/openthread/platform-abstraction/posix/posix_vendor_rcp.cmake -DOT_POSIX_CONFIG_RCP_VENDOR_INTERFACE=$GSDK_DIR/protocol/openthread/platform-abstraction/posix/cpc_interface.cpp -DOT_CLI_VENDOR_EXTENSION=$GSDK_DIR/protocol/openthread/platform-abstraction/posix/posix_vendor_cli.cmake -DOT_PLATFORM_CONFIG=openthread-core-silabs-posix-config.h

**Reply by suveshpratapa** [reply]
@dengyuxin2 Did you try tuning `OPENTHREAD_CONFIG_MAC_CSL_REQUEST_AHEAD_US` for the OTBR as suggested in the Application note section "**3.2.4 OTBR Configuration for CSL**" in my comment?

Do you mind attaching packet traces of the (working) FTD-SSED attachment as well?  I should already have the failing trace the OTBR-SSED setup.

**Reply by dengyuxin2** [reply]
@suveshpratapa  yes！I have set the value to 5000
![image](https://github.com/openthread/openthread/assets/158288585/b66b62b7-18ba-4e59-b6eb-4827b5765786)

Through the packet capture analysis, the child id response on the otbr side was not sent. The child id response packet was not seen in the packet capture. The otbr log showed that it was sent and no error was reported. What is the reason?
![image](https://github.com/openthread/openthread/assets/158288585/14907ba5-8ac0-4eb2-818d-7b8b5b71a35e)
![1713924879971](https://github.com/openthread/openthread/assets/158288585/09861ab2-fb83-4bff-aeb2-4dd3107477e3)

thread info：
Active Timestamp: 1
Channel: 21
Channel Mask: 0x07fff800
Ext PAN ID: 88bd74ffac739149
Mesh Local Prefix: fd33:25c9:ce10:6951::/64
Network Key: 76cf98c4a2b5ed39f755258b8baec6b3
Network Name: OpenThread-f121
PAN ID: 0xf121
PSKc: fac7f781dd358b4c52d78452f5cc00e2

the log and packet：

[log+packet.zip](https://github.com/openthread/openthread/files/15085353/log%2Bpacket.zip)

**Reply by suveshpratapa** [reply]
@dengyuxin2 

1. I don't see output showing the transmitted bytes, like `otbr-agent[12039]: 00:06:29.192 [D] Mac-----------: ==============================[TX len=117]==============================
`, when the the "Child ID Response" was supposedly sent by the OTBR.  It seems the tx started but for some reason was never sent.  I see you have debug logs on but there's no indication of where the packet was dropped.

2. Since it is a sleepy device, the child is supposed to send a child ID request, and then poll the parent for the child ID response.  I see data polls from the child, but not the child ID response from the parent.

This piece of exchange (child ID request - child ID response) has nothing to do with CSL, and is a pretty basic step of Thread attachment -- I'm concerned it is not working for you.

Couple of requests

1. Can you retry the test with OTBR configured with `-DOT_CHILD_SUPERVISION=OFF`?  This is not an option that is exposed anymore in the Thread stack, I'm just curious if you notice a change in behavior.
2. Do you happen to know how the sleepy device is configured, and what params if any you're setting for CSL params? Are you setting a polling period?
3. Do you mind attaching a trace showing passing attach? FED-SSED?

**Reply by dengyuxin2** [reply]
@suveshpratapa 
1、 This OT_CHILD_SUPERVISION option has been canceled in the new version and is forced on by default. set(OT_CHILD_SUPERVISION ON CACHE STRING "enable child supervision" FORCE)
2、It is impossible to know the parameters of third-party devices purchased online. I used matter's lock_app example to simulate the ICD sed device on your mg24 BRD4187C development board. The same situation applies. The parameters you mentioned should all be standard defaults. So it has nothing to do with csl features
![image](https://github.com/openthread/openthread/assets/158288585/329302b8-f431-40ee-a57a-097bb7debbec)

3、yes！I captured the packet and analyzed it. I first used chip-tool to add a lock_app icd sed device but failed to join the thread network. Then I used chip-tool to add an ftd device. After the ftd device was added successfully, I added the lock_app icd sed device. It can be added successfully, but it is a sub-device under ftd, not under otbr. When I power off the ftd device, the sed device should find a new parent node and reattach. Through packet capture analysis, I still haven't seen the child ID response sent by otbr to the child device, so I think the reason is that the child ID response of otbr was not sent out, but the specific reason is not very clear.
ftd-sed successful packet capture：
![image](https://github.com/openthread/openthread/assets/158288585/18efcd21-dae6-4882-a12a-a72e81e743bb)
shut down ftd  sed reattach packet capture:
![image](https://github.com/openthread/openthread/assets/158288585/5576a814-6cdd-44e3-8273-af61eb95464e)
 
thread info：
Active Timestamp: 1
Channel: 21
Channel Mask: 0x07fff800
Ext PAN ID: 4cc1fcc8323aaef9
Mesh Local Prefix: fd1f:36a5:e9e8:7c22::/64
Network Key: b8f8b97acbb193b302055e9b9281df04
Network Name: OpenThread-a9f4
PAN ID: 0xa9f4
PSKc: 0bdf6ed772375aedad5a8dbaf8e993ab
Security Policy: 672 onrc 0

the log and packet：
[ftd&sed-log+packet.zip](https://github.com/openthread/openthread/files/15108611/ftd.sed-log%2Bpacket.zip)

**Reply by suveshpratapa** [reply]
@dengyuxin2 

> This OT_CHILD_SUPERVISION option has been canceled in the new version and is forced on by default.

I was suggesting it because I saw you turn it explicitly on in the CMake steps above. I'm surprised you didn't see a build error for the option that is no longer available.

We'll need to analyze this further, I'm concerned that a basic attach sequence fails with no Child ID response. That being said, it is interesting that the successful trace (first screenshot above) is also showing no child ID response for the successful SED attach...

**Reply by dengyuxin2** [reply]
@suveshpratapa  
It is possible that the packet capture tool may have leaked packets.
After the sed device is powered off and powered on for a while, I can capture the child id response sent by the ftd device to the child device, but until now I have never captured the child id response packet sent by otbr.
![image](https://github.com/openthread/openthread/assets/158288585/f3036850-2708-4f09-a35e-3c2c1945eab5)

**Reply by silabs-gregoire** [reply]
Hi @dengyuxin2, for this issue can you please post this on the Silabs community? We can offer more support there as Silabs engineers actively review posts: https://community.silabs.com/

**Reply by suveshpratapa** [reply]
@dengyuxin2 On the community post, please include information about the third-party devices you're using if you can.

**Reply by dengyuxin2** [reply]
@silabs-gregoire @suveshpratapa I have posted this on the Silabs community, here is the link to ask the question  https://community.silabs.com/s/question/0D5Vm000009NGLSKA4/sed-device-always-fails-to-attach-to-otbr

**Comment by dengyuxin2** [comment]
@suveshpratapa @silabs-gregoire  hi! 
I tracked the code process of sending the child id response, and compared and analyzed the multi-protocol otbr logs of the official otbr and silabs on the Raspberry Pi. The response data of the low-power device will first be added to the child device information through this function IndirectSender::AddMessageForSleepyChild. , waiting for sending, waiting for the sub-device information to be received. The Get<Mac::Mac>().RequestIndirectFrameTransmission() function will be called in the DataPollHandler::HandleDataPoll function to send, but your otbr has never been called because of the received frame. It always shows that no ack has been received for confirmation.
Why does the received frame RxFrame::IsAckedWithFramePending function always return zero?
![image](https://github.com/openthread/openthread/assets/158288585/ecc04cfc-beda-408f-abeb-3e81ade286dd)
**silabs multi-protocol otbr log:**
The log shows that received frames ack-fp always show zero ack-fp=0
![image](https://github.com/openthread/openthread/assets/158288585/3580056b-157a-4aa8-9592-d1d3e6c5ff52)
**Raspberry Pi runs official otbr log:**
The log shows that ack-fd is sometimes 1, ack-fd=1, so the DataPollHandler::HandleDataPoll will call Get<Mac::Mac>().RequestIndirectFrameTransmission() to send the data. 
![image](https://github.com/openthread/openthread/assets/158288585/c855496f-9d38-481b-b90b-b7e0a493ebd4)

**Comment by Komorebi0525** [comment]
Hi, I have the same problem in SDK 4.3.3. Can you tell me if there is a solution for this problem? Thanks！ @suveshpratapa @silabs-gregoire

**Reply by suveshpratapa** [reply]
@Komorebi0525 I see you have already cross-posted it to the Silicon Labs community discussions above.  Unfortunately we don't have a solution on this GSDK version as we have a few important changes to the SDK made in future releases.

We had some trouble reproducing this issue.  Can you state whether your steps were exactly similar to the issue already posted here?

**Reply by dengyuxin2** [reply]
> Hi, I have the same problem in SDK 4.3.3. Can you tell me if there is a solution for this problem? Thanks！ @suveshpratapa @silabs-gregoire
We changed to the zigbee rcp + thread rcp solution, which can add low-power devices. The problem that zigbee ncp + thread rcp cannot add low-power
